src/server-logger.ts

```diff
@@ -0,0 +1,257 @@
+import request from 'superagent';
+import stringify from 'safe-json-stringify';
+import cloneDeep from 'lodash.clonedeep';
+
+import { Logger } from './logger';
+import { IDetail, ILoggerConfig, LogLevel, ITrace, ILogBufferItem, ILogMessage } from './interfaces';
+import { calculateLogMessageSize } from './utils';
+
+const MAX_LOG_SIZE = 14500;
+const DEFAULT_UPLOAD_DEBOUNCE = 4000;
+
+export class ServerLogger {
+  private config: ILoggerConfig;
+  private isInitialized: boolean = false;
+  private logger: Logger;
+  private logBuffer: Array<ILogBufferItem> = [];
+  private pendingHttpRequest: boolean = false;
+  private debounceLogUploadTime: number;
+  private debounceTimer: any = null;
+
+
+  constructor (config: ILoggerConfig, logger: Logger) {
+    this.config = config;
+    this.logger = logger;
+    /* if we have all the needed config options, set this up */
+    if (
+      !config.url ||
+      !config.accessToken ||
+      !config.appVersion
+    ) {
+      const errMessage = 'Lacking necessary config options to set up server logging. ' +
+        'Not sending logs to server for this logger instance';
+      this.logger.error(errMessage, { ...config }, true);
+      throw new Error(errMessage);
+    }
+
+    this.isInitialized = true;
+    this.debounceLogUploadTime = config.uploadDebounceTime || DEFAULT_UPLOAD_DEBOUNCE;
+
+    window.addEventListener('unload', this.sendAllLogsInstantly.bind(this));
+  }
+
+  public addLogToSend (logLevel: LogLevel, message: string | Error, details: IDetail | IDetail[]): void {
+    if (!this.isInitialized) {
+      return;
+    }
+
+    if (message instanceof Error) {
+      details = details || message;
+      message = message.message;
+    }
+
+    let logMessage = this.convertToLogMessage(message, details);
+    let trace = this.convertToTrace(logLevel, logMessage);
+    let traceMessageSize = calculateLogMessageSize(trace);
+
+    /* if the individual message exceeds the max allowed size, truncate it */
+    if (traceMessageSize > MAX_LOG_SIZE) {
+      const newTrace = this.truncateLog(logLevel, logMessage);
+      this.logger.warn('log message too large to send to server. truncating log details and/or message', {
```

Comment:
looks like these types of messages are already logged from the `truncateLog` fn.